### PR TITLE
Custom completer example

### PR DESCRIPTION
**Pull request by ohrely**
Adds a basic example for customizing notebook completer functionality.

Inadvertently addresses https://github.com/jupyterlab/extension-examples/issues/156 by disabling the notebooks plugin.

### CODE REVIEW

**Review by fcollonval — CHANGES_REQUESTED** [review]
Thanks for this great example @ohrely 

Some general comment, could you upgrade the example to use the latest cookiecutter template?
You can achieve then thanks to  `scripts/upgrade_extensions.py`

Could you add a note in the readme on the disabling cotr extension feature you are using?

For the tests, there is a big in the ci for the other example. For yours, you need to have some configuration files matching the one in hello world - it should be ok one you upgrade all extensions. And you need to insert the code snippets in the readme using `jlpm run embedme` at the root examples folder.

**Comment by ohrely** [comment]
> Some general comment, could you upgrade the example to use the latest cookiecutter template?
> You can achieve that thanks to `scripts/upgrade_extensions.py`

@fcollonval When I run this script, it makes modifications to setup.py and package.json in several of the example extensions and adds RELEASE.md to every extension.  Is this the correct behavior?

**Comment by fcollonval** [comment]
>  Is this the correct behavior?

Yes this is expected as new commits have been pushed to the template project recently. If you can ignore those in the other extensions, it will be appreciated (so this PR keeps focusing on the new example alone).

**Comment by ohrely** [comment]
The two build_all tests are failing because one inserted code snippet is getting "corrected" by pre-commit linting, so it no longer matches the (more indented) source of the snippet.  Is there a preferred way to address this?

I'm not sure what to make of the three failing build_extensions checks.

**Comment by fcollonval** [comment]
Thanks for pushing forward

To fix the linter error in the Readme, you should wrap the code snippets between special comments; e.g.

```
<!-- prettier-ignore-start -->
\`\`\`ts // src/index.ts#L12-L14 

activate: (app: JupyterFrontEnd) => { 
console.log('the JupyterLab main application:', app); 
}, 
\`\`\` 
<!-- prettier-ignore-end -->
```

Regarding the remaining error in the new example, you need to overwrite tsconfig.json, .eslintrc.js, .eslintignore and .gitignore with those in the hello-world example (this is to ensure homogenous config).

File: completer/README.md [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+3.  `CustomConnector`, a lightweight source of mocked completion results.
+
+The first part is contained in the `index.ts` file, the second is in `connector.ts`, and the third is in `customconnector.ts`.
+
+## Creating a custom DataConnector
+
+In `src/customconnector.ts` we define a `CustomConnector` to generate mock autocomplete suggestions. Like the `ContextConnector` it is based on, `CustomConnector` extends _JupyterLab_'s [`DataConnector`](https://jupyterlab.readthedocs.io/en/latest/api/classes/statedb.dataconnector.html) class.
+
+- Try adding a new token to `tokenList` in the `completionHint` function.
+- Try modifying the `completionHint` function to return no matches when `token.value.length` is less than 3.
```
Comment (fcollonval) [inline completer/README.md:33]: Some sentence before this, like `This custom connector will behave like this:`

Could you change this part to describe the actual code doing those actions?

**Review by fcollonval — CHANGES_REQUESTED** [review]
Thanks @ohrely I push a commit to fix all linter troubles to go forward. I have two comments in the code and one general one: could you use `you` instead of `we` in the readme?

**Comment by hbcarlos** [comment]
Hi @ohrely and @fcollonval, I rebased from master and added a basic test.

I'll update the extension to JupyterLab v3.1 too.

**Comment by fcollonval** [comment]
@hbcarlos thanks for pushing this one. Could you please add a more advanced ui test that open a notebook, trigger the completer and check the suggestions are the one expected?

**Comment by hbcarlos** [comment]
@fcollonval I'm not sure where the CI test fail. Do you have any advice?

**Comment by fcollonval** [comment]
We are getting closer 🙃

The integration test needs to be more robust. Did you try to run it locally before running it in the ci?
Two ideas, the kernel may not be ready when you hit tab. So waiting for the idle state could be good. The other one is that I had some trouble with codemirror edition that I circumvent by using `page.keyboard` - see settings example.

**Comment by hbcarlos** [comment]
Thanks for the advice, @fcollonval! Yes, I tested locally before pushing to CI.

**Comment by hbcarlos** [comment]
Btw, when running locally, If you don't run `npx playwright install` first, the test fails because chromium is not installed.

**Comment by fcollonval** [comment]
Thanks @hbcarlos 

I updated the test to trigger the custom completion and make it more robust.
Plus I catched some missing package not listed in dependencies.

**Review by fcollonval — APPROVED** [review]
Let's merge
Thanks a lot @ohrely and @hbcarlos

**Comment by hbcarlos** [comment]
Thanks to you for all the help!!

**Comment by ohrely** [comment]
Thank you both!

**Pull request merged**